Design exploration: structured metadata two ways

sryza opened the pull request:
This presents an approach to structured asset metadata, with two variants: one nested and one non-nested.

Nowadays, you can do this:
```python
return MaterializeResult(
    metadata={"row_count": 2309, "source_code_link" "http://...", "columns": my_column_schema}
)
```

Two problems with this:
- There are no corresponding symbols in the Python API. Python symbols act as documentation for the metadata, allow it to be type-checked, and allow the use of auto-complete.
- It's not namespaced. If I have an asset that contains data about a farm, and I come across a metadata key named "row_count", how do I know whether it's referring to "the number of rows in a warehouse table" or "the number of rows of cabbages on the farm"?

The approached proposed here looks somewhat similar to the "Approach 1" that @rexledesma described [in this Notion doc](https://www.notion.so/dagster/Standardizing-metadata-fields-7efda2c4de754717a026bc00601f6b84).

### Common to both of the variants:
- Does not change or add any properties on `MaterializeResult` or `AssetSpec`. I.e. the structured metadata classes involved are intended to be translated to the plain dictionaries that these APIs already expect.
- Structured metadata can be intermixed with unstructured metadata.
- Uses Python package as a namespacing convention. If you encounter a metadata entry on asset "foo" with namespace="package_x", key="row_count", and value="7", you're meant to understand that "foo's row_count is 7, according to the definition of row_count intended by the developers of package_x".
- All the subclasses of `NamespacedMetadata` are identical across both examples.
- They maintain our ability to interpret a metadata dict as a flat list of metadata entries, for display in the UI, like this:
<img width="684" alt="image" src="https://github.com/dagster-io/dagster/assets/654855/e5b525cf-32b7-44f1-a7c3-c4849e488de5">

### Differences between the variants
- The constructed metadata dictionaries are different between the variants.
  - In `standardized-metadata-flexible-spec.py`, the namespaces are included as parts of the metadata keys.
  - In `standardized-metadata-nested.py`, the metadata attributes are nested underneath their namespaces, which are themselves nested underneath a `__namespaced_metadata` top-level key.
- The implementation of `asset2`, which showcases the constructed metadata dictionaries, is different between the variants.

### Comparing the approaches

In favor of the unnested approach:

- In the unnested approach, each entry in the dictionary that's passed to the `metadata` argument corresponds to a metadata entry shown in the UI. This feels simple and expected, and is consistent way the way Dagster metadata currently works.
- Consequently, basic operations like checking whether a metadata dictionary contains an expected entry are very straightforward and don't require use of helper APIs to interpret the metadata.
- The unnested approach matches our convention for naming keys in tag dictionaries, where we use namespace prefixes like "dagster/" and "dagster-kubernetes/".
- The nested approach requires introducing an additional API for constructing the metadata dictionary –
 `build_metadata_dict`.
- In the unnested approach, you can use basic Python dictionary manipulation to combine metadata entries from different sources. Users don't need to learn the behavior bespoke APIs for combining metadata.

In favor of the nested approach:

- In the unnested approach, if we want the UI to render metadata entries in a way that gives special treatment to the namespace, then we need to rely on string parsing to separate it out.
- If there are metadata keys that include slashes, but the part before the slash isn't intended to be the namespace, then any special treatment we apply in the UI will still treat it as the namespace, which will look weird.
- Even if we set the namespace convention going forward, there's still historical metadata keys from before this was a convention.

### Relevant other systems
- [OpenLineage Facets](https://openlineage.io/docs/spec/facets/).
  - Individual OpenLineage facets roughly correspond to Dagster metadata entries. Example built-in OpenLineage facets include "columnLineage", "ownership", "schema", "storage", and "outputStatistics" (which is just rowCount and size).
  - Facets keys are not namespaced, but there's a naming convention for non-built-in facets that includes the name of the technology they relate to
  - Facet values are schematized JSON
- [Kubernetes Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
  - Here's their documentation on "well-known" annotations: https://kubernetes.io/docs/reference/labels-annotations-taints/
  - They use slashes to delimit within annotation keys, e.g. "apf.kubernetes.io/autoupdate-spec"
  - Annotation values are just strings.

schrockn commented on standardized-metadata-nested.py:
```diff
@@ -0,0 +1,225 @@
+@asset
+def asset2():
+    return MaterializeResult(
+        metadata={
+            "dagster.table": {
+                "_version": "1.0.0",
+                "num_rows_inserted": 5,
+                "columns": [{"name": "user_id", "data_type": "str"}],
+                "num_rows_total": 500,
+            },
+            "dagster": {
+                "_version": "1.0.0",
+                "storage_kind": "snowflake",
+                "storage_address_string": "my_db.my_schema.asset1",
+            },
+            "dagster_snowflake": {
+                "_version": "1.0.0",
+                "snowflake_address": {
+                    "database": "my_db",
+                    "db_schema": "my_schema",
+                    "table_name": "asset1",
+                },
+            },
+            # non-structured metadata
+            "my_unschematized_piece_of_metadata": 5,
+            "staging_address": {
+                "database": "my_staging_db",
+                "db_schema": "my_schema",
+                "table_name": "asset1",
+            },
+        }
+    )
```
Definitely prefer this to the special treatment of the top-level keys in option one. I think there are two issues here:

1) There is no schema to avoid naming collisions with top-level fields past, present, and future. I still like we should hide of all of these underneath some sort of special key.
2). Do we want to be able to place multiple instances of metadata blobs? This would mean the option to "name" them.

sryza replied:
> Do we want to be able to place multiple instances of metadata blobs? This would mean the option to "name" them.

Thinking aloud here:

To make this concrete, let's imagine a situation where someone would want the Snowflake "cluster_by" key ("cluster_by" is an attribute of a Snowflake table) to show up multiple times within the metadata blob for a single AssetMaterialization.

So to back up, why would they want to do this? The main reason I can think of would be because there are multiple Snowflake tables that are relevant to a particular materialization. For example, there's the Snowflake table that's the asset itself that's being materialized, and then there's also a "staging" Snowflake table that was somehow produced during the process of materializing the asset.

So assuming we want to record some Snowflake-specific attributes about this staging table, it's possible that we'd also want to log some other metadata fields that are relevant to tables, but not just Snowflake tables, e.g. "num_rows_total".

So, at least naively, I would imagine the materialization-level metadata blob should end up containing some sub-blob that represents this staging table and contains the set of attributes that describe it.

In the unnested variant, this would look like:
```json
{
    "dagster_snowflake/cluster_by": "user_name",
    "staging_table": {
        "dagster_snowflake/cluster_by": "user_id",
        "dagster.table/num_rows_total": 5
    }
}
```

In the nested variant, this would look like:
```json
{
    "__namespaced_metadata": {
        "dagster_snowflake": {"cluster_by": "user_id"}
    }
    "staging_table": {
        "__namespaced_metadata": {
            "dagster_snowflake": {"cluster_by": "user_id"},
            "dagster.table": {"num_rows_total": 5}
        }
    }
}
```

sryza commented:
I realized an issue with the original nested variant proposed here: it was impossible to interpret the metadata dict as a table of metadata entry, like we support in our UI. This is because it was impossible to distinguish between "namespaces" and "metadata entry" at the top level.
![image](https://github.com/dagster-io/dagster/assets/654855/4e892381-db81-42bc-b1e1-6960392d404e)

I addressed this by adding a reserved top-level "__namespaced_metadata" key that all the namespaced metadata goes underneath.

So the nested metadata blob now looks like:

```json
{
    "__namespaced_metadata": {
        "dagster.table": {
            "_version": "1.0.0",
            "num_rows_inserted": 5,
            "columns": [{"name": "user_id", "data_type": "str"}],
            "num_rows_total": 500,
        },
        "dagster": {
            "_version": "1.0.0",
            "storage_kind": "snowflake",
            "storage_address_string": "my_db.my_schema.asset1",
        },
        "dagster_snowflake": {
            "_version": "1.0.0",
            "snowflake_address": {
                "database": "my_db",
                "db_schema": "my_schema",
                "table_name": "asset1",
            },
        },
    },
    "my_unschematized_piece_of_metadata": 5,
    "staging_address": {
        "database": "my_staging_db",
        "db_schema": "my_schema",
        "table_name": "asset1",
    },
}
```

A consequence of this is that it's no longer possible to construct the metadata dict using splatting. Instead, I introduced a `build_metadata_dict` function:

```python
    return MaterializeResult(
        metadata=build_metadata_dict(
            namespaced_metadatas=[
                AssetMetadata(
                    storage_kind="snowflake", storage_address_string="my_db.my_schema.asset1"
                ),
                TableMaterializationMetadata(
                    num_rows_total=500,
                    num_rows_inserted=5,
                    columns=[TableColumn(name="user_id", data_type="str")],
                ),
                SnowflakeTableMetadata(
                    snowflake_address=SnowflakeTableAddress(
                        database="my_db", db_schema="my_schema", table_name="asset1"
                    )
                ),
            ],
            metadata={
                "my_unschematized_piece_of_metadata": 5,
                "staging_address": SnowflakeTableAddress(
                    database="my_staging_db", db_schema="my_schema", table_name="asset1"
                ).dict(),
            },
        )
    )
```

We could potentially eliminate this function by adding a `namespaced_metadata` argument to all APIs that currently accept `metadata`.

sryza commented:
@schrockn I pushed an update that adds in a top-level `__namespaced_metadata` key to the nested approach.  More detail here: https://github.com/dagster-io/dagster/pull/20077#issuecomment-1967955244.

A followup thought: we make heavy use of the "dagster/" prefix in our tag keys.  If we moved towards the nested approach for metadata, would we want to do the same for tag dicts as well?

schrockn commented on standardized-metadata-nested.py:
```diff
@@ -0,0 +1,259 @@
+def asset2():
+    return MaterializeResult(
+        metadata={
+            "__namespaced_metadata": {
+                "dagster.table": {
+                    "_version": "1.0.0",
+                    "num_rows_inserted": 5,
+                    "columns": [{"name": "user_id", "data_type": "str"}],
+                    "num_rows_total": 500,
+                },
+                "dagster": {
+                    "_version": "1.0.0",
+                    "storage_kind": "snowflake",
+                    "storage_address_string": "my_db.my_schema.asset1",
+                },
+                "dagster_snowflake": {
+                    "_version": "1.0.0",
+                    "snowflake_address": {
+                        "database": "my_db",
+                        "db_schema": "my_schema",
+                        "table_name": "asset1",
+                    },
+                },
+            },
```
This makes a ton more sense to me. I can understand how this system works just by looking at this serialization format. 👍🏻

schrockn commented:
> A followup thought: we make heavy use of the "dagster/" prefix in our tag keys. If we moved towards the nested approach for metadata, would we want to do the same for tag dicts as well?

I don't think so. I think tags should remain simple key/value pairs so that search UI, indexes, and other constructs can trivially deal with them. Nested, complex data types should be in the province of metadata only.

PedramNavid commented on standardized-metadata-flexible-spec.py:
```diff
@@ -0,0 +1,219 @@
+        metadata={
+            "dagster.table/num_rows_inserted": 5,
+            "dagster.table/columns": [{"name": "user_id", "data_type": "str"}],
+            "dagster.table/num_rows_total": 500,
```
what happens if i try to emit metadata here that isn't part of the class? e.g. dagster.table/name: foo

sryza replied:
We can't stop you from doing this (in either of the variants). However, you could run something like `TableMaterializationMetadata.validate(metadata)` to assert that you're yielding known entries.

PedramNavid replied:
But the metadata will still make it through? I think that's fine, better than an error.

sryza replied:
Right, the metadata will still make it through

PedramNavid commented:
Overall this makes sense but I do wonder if there's an opportunity to tease out two types of metadata we're using for two very different purposes here. 

It seems we have what I would consider both metrics and labels being used as metadata. If we plan on changing how we structure metadata, what if we went a little further? Something like: 

```yaml
labels: 
  app: snowflake
  env: prod

metrics:
  - snowflake:
    - name: rows_inserted
      value: 500
    - name: partitions
      value: 5
  # Optionally
    - dagster:
        - storage_kind: snowflake
        - storage_address_string: "mydb.my_schema.asset1"
    - my_unschematized_piece_of_metadata: 5
    - staging_address: 
        database: my_staging_db
        db_schema: my_schema
        table_name: asset1
```

sryza commented:
> It seems we have what I would consider both metrics and labels being used as metadata. If we plan on changing how we structure metadata, what if we went a little further?

Current thinking is that we would use "tags" for what you're calling labels. I.e. tags would be for searching/filtering/categorizing things. "Metadata" would be for understanding those things more deeply once you've found them.

Here's a larger proposal for tags: https://github.com/dagster-io/internal/discussions/8544

Regarding the name "metric", a concern is that it might set up the expectation that the values would be numeric. E.g. would you consider {"storage_address_string": "mydb.my_schema.asset1"} a metric?

PedramNavid commented:
> Current thinking is that we would use "tags" for what you're calling labels. I.e. tags would be for searching/filtering/categorizing things. "Metadata" would be for understanding those things more deeply once you've found them.
> 
> Here's a larger proposal for tags: [dagster-io/internal#8544](https://github.com/dagster-io/internal/discussions/8544)
> 
> Regarding the name "metric", a concern is that it might set up the expectation that the values would be numeric. E.g. would you consider {"storage_address_string": "mydb.my_schema.asset1"} a metric?

I think I'd consider that a label? It's something I might want to search for, but it is likely to be the same for the asset over time (almost by definition in this case). I realize I broke my example above with a bad copy paste, something like this is what I meant: 

```yaml
labels: 
  app: snowflake
  env: prod

# Optionally
labels:
  - dagster:
        - storage_kind: snowflake
        - storage_address_string: "mydb.my_schema.asset1"
    - staging_address: 
        database: my_staging_db
        db_schema: my_schema
        table_name: asset1

metrics:
  - default: 
     - name: execution_time
       value: 50
  - snowflake: # namespace 
    - name: rows_inserted
      value: 500
    - name: partitions
      value: 5

# Optionally
metrics:
    - my_unschematized_piece_of_metadata: 5
```

Metrics then become things we plot, aggregate, use for anomalies, etc.
Labels are dimensions which we can use for filtering, grouping, categorizing.  But it looks like tags and labels are one-and-the-same? As long as they can be used for insights as well as search/filter then I think that works just as well.

sryza commented:
> I think I'd consider that a label? It's something I might want to search for, but it is likely to be the same for the asset over time (almost by definition in this case). I realize I broke my example above with a bad copy paste, something like this is what I meant:

@PedramNavid got it, that makes sense.

While not pictured in my example above, I think there will be metadata values that aren't numeric but also don't make sense for searching and filtering.  For example, a JSON blob that encodes the list of all column names, types, and descriptions.  Or a similar JSON blob that encodes column lineage.

sryza commented:
Feedback from @hellendag on the unnested approach is that it sounds fine if (and only if) there's a rule we can follow reliably for parsing.

schrockn commented:
@sryza and I discussed this and I'm supportive of going through what I am calling the "metadata mix-in" approach.

My one request is that we do the exercise is modeling this all the way through GraphQL to the frontend so we agree on a full end-to-end mental model. That way we can develop a common language across the whole stack for this "metadata envelope" or whatever we will call it. cc: @hellendag

schrockn commented:
> When you say "metadata envelope", I'm assuming you mean a set of metadata entries with keys in the same namespace, with namespaces roughly corresponding to Python packages, e.g. "dagster", "dagster.table", "dagster-snowflake".

Yes. Corresponding to an object in Python like `SnowflakeTableMetadata`.
